File: src/main/scala/com/nibado/projects/adventscala/Day08.scala
package com.nibado.projects.adventscala

class Day08 extends Day {
  override def run(): Unit = {
    val input = getResource("/day08.txt")

    printAnswer(8, "One", input.map(s => s.length - unescape(s).length).sum)
    printAnswer(8, "Two", input.map(s => escape(s).length - s.length).sum)
  }

  def unescape(input: String): String = {
    val it = input.substring(1, input.length - 1).iterator
    val out = new StringBuilder
    while(it.hasNext) {
      val c = it.next()

      if(c == '\\') {
        val c2 = it.next()
        if(c2 == '"' || c2 == '\\') {
          out.append(c2)
        }
        else if(c2 == 'x') {
          val v = Integer.parseInt(it.next().toString + it.next.toString, 16)
          out.append(v.toChar)
        }
      }
      else {
        out.append(c)
      }
    }

    out.toString()
  }

  def escape(input: String): String = {
    '"' + input.flatMap(c => {
      if(c == '"' || c == '\\') List('\\', c) else List(c)
    }) + '"'
  }
}

File: src/main/java/com/nibado/projects/advent/y2015/Day08.java
package com.nibado.projects.advent.y2015;

import java.util.List;
import java.util.stream.Collectors;

import static com.nibado.projects.advent.Util.printAnswer;
import static com.nibado.projects.advent.Util.readResource;

public class Day08 implements Runnable {
    @Override
    public void run() {
        List<String> list = readResource("/2015/day08.txt");

        int result1 = list.stream().collect(Collectors.summingInt(s -> s.length() - (unescape(s).length())));
        printAnswer(result1);

        int result2 = list.stream().collect(Collectors.summingInt(s -> escape(s).length() - s.length()));
        printAnswer(result2);
    }

    public String unescape(String input) {
        StringBuilder out = new StringBuilder(input.length());

        for(int i = 1;i < input.length() - 1;i++) {
            char c = input.charAt(i);
            if(c == '\\') {
                i++;
                char c2 = input.charAt(i);
                if(c2 == '\\' || c2 == '"') {
                    out.append(c2);
                }
                else if(c2 == 'x') {
                    i++;
                    c2 = (char)Integer.parseInt(input.substring(i, i + 2), 16);
                    out.append(c2);
                    i += 1;
                }
            }
            else {
                out.append(c);
            }
        }

        return out.toString();
    }

    public String escape(String input) {
        StringBuilder out = new StringBuilder(input.length() * 2);
        out.append('"');
        for(char c : input.toCharArray()) {
            if(c == '"' || c == '\\') {
                out.append('\\');
            }
            out.append(c);
        }

        out.append('"');

        return out.toString();
    }

    public static void main(String... argv) {
        new Day08().run();
    }
}
